### PR TITLE
Add remaining missing types and disable `noImplicitAny`

### PR DESCRIPTION
**Pull request by robertknight**
Add remaining missing types so that the code typechecks with `noImplicitAny`
enabled in tsconfig.json.

### CODE REVIEW

File: src/background/hypothesis-chrome-extension.js [hunk 1]
```diff
@@ -150,20 +154,24 @@ export default function HypothesisChromeExtension({
   // exposed for use by tests
   this._onTabStateChange = onTabStateChange;
 
+  /** @param {chrome.tabs.Tab} tab */
   function onBrowserActionClicked(tab) {
-    const tabError = state.getState(tab.id).error;
+    const tabId = /** @type {number} */ (tab.id);
```
Comment (robertknight) [inline src/background/hypothesis-chrome-extension.js:159]: I am assuming in several places in this PR that tabs have an ID and URL. We've never seen any reports of errors due to this not being the case, though I couldn't completely guarantee that it has never happened.

File: src/background/settings.js [hunk 2]
```diff
@@ -1,7 +1,16 @@
+/**
+ * Incomplete type for settings in the `settings.json` file.
```
Comment (robertknight) [inline src/background/settings.js:2]: The intent is that we'll fill this out with additional fields as we need to.

File: src/background/browser-action.js [hunk 3]
```diff
@@ -17,8 +17,12 @@ const icons = {
   },
 };
 
-// themes to apply to the toolbar icon badge depending on the type of
-// build. Production builds use the default color and no text
+/**
+ * Themes to apply to the toolbar icon badge depending on the type of
+ * build. Production builds use the default color and no text
+ *
+ * @type {Record<string, { defaultText: string, color: string }>}
```
Comment (esanzgar) [inline src/background/browser-action.js:24]:
```suggestion
 * @type {Record<string, {defaultText: string, color: string}>}
```

File: src/background/tab-state.js [hunk 4]
```diff
@@ -77,14 +79,14 @@ export default function TabState(initialState, onchange) {
...
-   *                   The provided state will be merged with the default
-   *                   state for a tab.
+   * @param {Record<number, Partial<State>>} newState -
+   *   Map of tab ID to state. The state is merged with the default state for a tab.
    */
   this.load = function (newState) {
     /** @type {TabStateMap} */
     const newCurrentState = {};
-    Object.keys(newState).forEach(function (tabId) {
+    Object.keys(newState).forEach(function (tabIdStr) {
```
Comment (esanzgar) [inline src/background/tab-state.js:88]:
```suggestion
    Object.keys(newState).forEach(tabIdStr => {
```

As part of this modernisation, we could run `eslint` to automatically convert to arrow functions, when possible.

**Review by esanzgar — APPROVED** [review]
LGTM

File: src/background/browser-action.js (hunk 3 above)
Comment (robertknight) [reply src/background/browser-action.js:24]: There aren't a ton of examples but we do tend to include a space before and after parens in object types. The general rule of thumb I follow is to format types as Prettier would (in a TypeScript file, unfortunately it doesn't format JSDoc types).

File: src/background/tab-state.js (hunk 4 above)
Comment (robertknight) [reply src/background/tab-state.js:88]: Yes. This was on my list to do at some point.

**Pull request merged**